sklearn/decomposition/tests/test_pca.py

```diff
@@ -589,3 +589,35 @@ def test_pca_bad_solver():
     X = np.random.RandomState(0).rand(5, 4)
     pca = PCA(n_components=3, svd_solver='bad_argument')
     assert_raises(ValueError, pca.fit, X)
+
+
+def test_pca_dtype_preservation():
+    for svd_solver in solver_list:
+        # Ensure that PCA does not upscale the dtype when input is float32
+        X_64 = np.random.RandomState(0).rand(1000, 4).astype(np.float64)
+        X_32 = X_64.astype(np.float32)
+
+        pca_64 = PCA(n_components=3, svd_solver=svd_solver,
+                     random_state=0).fit(X_64)
+        pca_32 = PCA(n_components=3, svd_solver=svd_solver,
+                     random_state=0).fit(X_32)
+
+        assert pca_64.components_.dtype == np.float64
+        assert pca_32.components_.dtype == np.float32
+        assert_array_almost_equal(pca_64.components_, pca_32.components_,
+                                  decimal=5)
```

Comment:
Please also check the dtype preservation for:

```python
assert pca_64.transform(X_64) == np.float64
assert pca_32.transform(X_32) == np.float32
```

This is a straightforward consequence of the above but better be explicit :)